addons/account/models/partner.py

```diff
@@ -434,6 +434,22 @@ def _compute_bank_count(self):
         for partner in self:
             partner.bank_account_count = mapped_data.get(partner.id, 0)
 
+    @api.constrains('company_id')
+    def _check_company_and_invoices(self):
+        partners_with_company = self.filtered(lambda p: p.company_id)
+        if partners_with_company:
+            query = """SELECT p.id
+                    FROM res_partner as p
+                    WHERE p.id in %s
+                        AND EXISTS (SELECT 1
+                                    FROM account_invoice as inv
+                                    WHERE inv.partner_id = p.id
+                                        AND p.company_id <> inv.company_id)
+                    LIMIT 1"""
```

Comment:
I would expect this to be faster; and easier to read
```suggestion
            query = """
                SELECT partner.id
                  FROM res_partner partner
                  JOIN account_invoice invoice ON invoice.partner_id = partner.id
                                              AND invoice.company_id != partner.company_id
                 WHERE partner.id IN %s
                 LIMIT 1
             """
```